pkg/cloud/aws/actuators/machine_scope.go

```diff
@@ -131,3 +134,46 @@ func (m *MachineScope) Close() {
 		klog.Errorf("[machinescope] failed to store provider status for machine %q in namespace %q: %v", m.Machine.Name, m.Machine.Namespace, err)
 	}
 }
+
+func machineConfigFromProviderSpec(clusterClient client.MachineClassesGetter, providerConfig clusterv1.ProviderSpec) (*v1alpha1.AWSMachineProviderSpec, error) {
+	var config v1alpha1.AWSMachineProviderSpec
+	if providerConfig.Value != nil {
+		klog.V(4).Info("Decoding ProviderConfig from Value")
+		return unmarshalProviderSpec(providerConfig.Value)
+	}
+
+	if providerConfig.ValueFrom != nil && providerConfig.ValueFrom.MachineClass != nil {
+		ref := providerConfig.ValueFrom.MachineClass
+		klog.V(4).Info("Decoding ProviderConfig from MachineClass")
+		if ref != nil {
+			klog.V(6).Infof("ref: %v", ref)
+			if ref.Provider != "" && ref.Provider != "aws" {
```

Comment:
does this imply that the default provider is "aws" when `ref.Provider` is empty?

Reply:
I figured we should also attempt to decode the providerSpec if the user doesn't specify a provider in the MachineClass, so instead of returning an error here the error will be returned when attempting to unmarshal the providerSpec if it is meant for a different provider.